arch/arm/src/s32k1xx/s32k1xx_flexio_i2c.c

```diff
@@ -0,0 +1,755 @@
+{
+  struct s32k1xx_flexio_i2cdev_s *priv;
+
+  if (port == 0)
+    {
+      priv          = &g_i2c0dev;
+      priv->dev.ops = &s32k1xx_flexio_i2c_ops;
+    }
+  else
+    {
+      i2cerr("I2C Only support 0\n");
+      return NULL;
+    }
+
+  nxmutex_lock(&priv->lock);
+
+  /* Test if already initialized or not */
+
+  if (1 < ++priv->refs)
+    {
+      nxmutex_unlock(&priv->lock);
+      return &priv->dev;
+    }
+
+  priv->port      = port;
+  priv->frequency = 0;
+
+  s32k1xx_get_pclkfreq(FLEXIO0_CLK, &priv->base_freq);
+
+  /* Configure pins */
+
+  s32k1xx_pinconfig(priv->config->scl_pin);
+  s32k1xx_pinconfig(priv->config->sda_pin);
+
+  s32k1xx_flexio_i2c_init(priv);
+
+  /* Attach Interrupt Handler */
+
+  irq_attach(priv->irqid, s32k1xx_flexio_i2c_interrupt, priv);
+
+  /* Enable Interrupt Handler */
+
+  up_enable_irq(priv->irqid);
+
+  nxmutex_unlock(&priv->lock);
+  return &priv->dev;
+}
+
+/****************************************************************************
+ * Name: s32k1xx_flexio_i2cbus_uninitialize
+ *
+ * Description:
+ *   Uninitialise an I2C device
+ *
+ ****************************************************************************/
+
+int s32k1xx_flexio_i2cbus_uninitialize(struct i2c_master_s *dev)
+{
+  struct s32k1xx_flexio_i2cdev_s *priv =
+              (struct s32k1xx_flexio_i2cdev_s *)dev;
+
+  /* Decrement reference count and check for underflow */
+
+  if (priv->refs == 0)
+    {
+      return ERROR;
+    }
+
+  nxmutex_lock(&priv->lock);
+  if (--priv->refs)
+    {
+      nxmutex_unlock(&priv->lock);
+      return OK;
+    }
+
+  up_disable_irq(priv->irqid);
+  irq_detach(priv->irqid);
+
+  nxmutex_unlock(&priv->lock);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: s32k1xx_flexio_i2c_transfer
+ *
+ * Description:
+ *   Perform a sequence of I2C transfers
+ *
+ * TODO: Multiple i2c_msg_s read operations with the same address are not
+ * currently guaranteed.
+ ****************************************************************************/
+
+static int s32k1xx_flexio_i2c_transfer(struct i2c_master_s *dev,
+                               struct i2c_msg_s *msgs, int count)
+{
+  struct s32k1xx_flexio_i2cdev_s *priv =
+              (struct s32k1xx_flexio_i2cdev_s *)dev;
+  int i;
+  int ret    = 0;
+  int semval = 0;
+  struct timespec abs_time;
+
+  if (msgs->frequency > I2C_MAX_FREQUENCY)
+    {
+      return -EINVAL;
+    }
+
+  DEBUGASSERT(dev != NULL);
+
+  /* Get exclusive access to the I2C bus */
+
+  nxmutex_lock(&priv->lock);
+
+  /* Check wait semaphore value. If the value is not 0, the transfer can not
+   * be performed normally.
+   */
+
+  ret = nxsem_get_value(&priv->wait, &semval);
+  DEBUGASSERT(ret == OK && semval == 0);
+
+  priv->total_bytes = 0;
+
+  for (i = 0; i < count; i++)
+    {
+      priv->total_bytes += msgs[i].length;
+    }
+
+  if (priv->total_bytes < 12)
```

Comment:
Suggestion include the comment here too:
/* FlexIO driver can only handle transfers of maximum 12 bytes */